### PR TITLE
Use context menu handler for callback

### PR DESCRIPTION
**Pull request by eyvallah**
only modified the qt version.

### CODE REVIEW

**Comment by eyvallah** [comment]
@375gnu I brought the change also to other platforms, I can't test it though because I'm on Mac only, the gtk one is apparently wrong, there needs to be some mechanism to pass the app argument.

File: src/celestia/celestiacore.h [hunk 1]
```diff
@@ -372,6 +368,16 @@ class CelestiaCore // : public Watchable<CelestiaCore>
     void setCursorHandler(CursorHandler*);
     CursorHandler* getCursorHandler() const;
 
+    class ContextMenuHandler
+    {
+    public:
+        virtual ~ContextMenuHandler() {};
```
Comment (375gnu) [inline src/celestia/celestiacore.h]: {} can be replaced with `= default`

File: src/celestia/gtk/menu-context.cpp [hunk 2]
```diff
@@ -25,33 +25,25 @@
...
-
-/* There is no way to pass the AppData struct to the menu at this time. This
- * keeps the global variable local to this file, at least conceptually. */
-static AppData* app;
-
-
-/* Initializer. Sets the appData, since there is no way to pass it. */
-void initContext(AppData* a)
+/* Initializer. Sets the appData */
+GTKContextMenuHandler::GTKContextMenuHandler(AppData* app) : CelestiaCore::ContextMenuHandler(), app(app)
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.cpp]: please change style:
```
GTKContextMenuHandler::GTKContextMenuHandler(AppData* _app) :
    CelestiaCore::ContextMenuHandler(),
    app(_app)
```

File: src/celestia/gtk/menu-context.h [hunk 3]
```diff
@@ -21,11 +21,14 @@
 
 #include "common.h"
 
+class GTKContextMenuHandler : public CelestiaCore::ContextMenuHandler {
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.h]: move { to the next line

File: src/celestia/gtk/menu-context.h [hunk 4]
```diff
@@ -21,11 +21,14 @@
...
+public:
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.h]: before public/private we use 1 space indent

File: src/celestia/kde/kdeapp.cpp [hunk 5]
```diff
@@ -106,7 +106,7 @@ static uint32 FilterOtherLocations = ~(Location::City |
                     Location::Vallis |
                     Location::Mare);
 
-KdeApp::KdeApp(std::string config, std::string dir, std::vector<std::string> extrasDirs, bool fullscreen, bool disableSplash) : KMainWindow(0, 0)
+KdeApp::KdeApp(std::string config, std::string dir, std::vector<std::string> extrasDirs, bool fullscreen, bool disableSplash) : KMainWindow(0, 0), CelestiaCore::ContextMenuHandler()
```
Comment (375gnu) [inline src/celestia/kde/kdeapp.cpp]: no need to change kde frontend, it's dead

File: src/celestia/qt/qtappwin.cpp [hunk 6]
```diff
@@ -182,7 +176,7 @@ void FPSActionGroup::updateFPS(int fps)
 }
 
 CelestiaAppWindow::CelestiaAppWindow(QWidget* parent) :
-    QMainWindow(parent)
+    QMainWindow(parent), CelestiaCore::ContextMenuHandler()
```
Comment (375gnu) [inline src/celestia/qt/qtappwin.cpp]: please move CelestiaCore::ContextMenuHandler() to the next line

File: src/celestia/win32/winmain.cpp [hunk 7]
```diff
@@ -2349,6 +2343,17 @@ class WinAlerter : public CelestiaCore::Alerter
     }
 };
 
+class WinContextMenuHandler : public CelestiaCore::ContextMenuHandler
+{
+public:
+    WinContextMenuHandler() : CelestiaCore::ContextMenuHandler() {};
```
Comment (375gnu) [inline src/celestia/win32/winmain.cpp]: i don't think that you need to define a constructor here

File: src/celestia/gtk/menu-context.h (hunk 4 above)
Comment (eyvallah) [reply src/celestia/gtk/menu-context.h]: i fixed this one but it seems our win32 code doesn't follow that

**Comment by eyvallah** [comment]
i think ill stick with the delegate class instead of the lambda for now, to match with alerter, cursor handler, watcher etc.

**Comment by 375gnu** [comment]
> the gtk one is apparently wrong,

yes, i tried to compile it and it threw a lot of errors, and i'm not able to fix them yet as C code doesn't map to C++ properly.

**Comment by eyvallah** [comment]
@375gnu falls back to the old way for gtk for now, can you verify if it passes compilation?

File: src/celestia/gtk/menu-context.h [hunk 8]
```diff
@@ -21,11 +21,15 @@
...
 
-/* Initializer Function */
-void initContext(AppData* a);
+    void requestContextMenu(float, float, Selection sel);
 
-/* Entry Function */
-void menuContext(float, float, Selection sel);
+ private:
+    AppData *appData;
+}
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.h]: missing ';'

File: src/celestia/gtk/menu-context.h [hunk 9]
```diff
@@ -21,11 +21,15 @@
...
+    GTKContextMenuHandler(AppData* app);
 
-/* Initializer Function */
-void initContext(AppData* a);
+    void requestContextMenu(float, float, Selection sel);
 
-/* Entry Function */
-void menuContext(float, float, Selection sel);
+ private:
+    AppData *appData;
```
Comment (375gnu) [inline src/celestia/gtk/menu-context.h]: it seems that now you don't need to store AppData * anymore

**Comment by pirogronian** [comment]
@375gnu I think class is just more general and flexible than function.

**Comment by eyvallah** [comment]
@375gnu yes, as pirogronian said, it's more flexible and it's in line with the other delegate/handler method in the celestiacore.

**Comment by 375gnu** [comment]
@eyvallah feel free to merge, we usually merge own pr ourselves

**Pull request merged**